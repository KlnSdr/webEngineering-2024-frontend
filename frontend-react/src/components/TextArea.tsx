import React, { useEffect, useState } from "react";
import Heading from "./Heading";

/**
 * Props for the TextArea component.
 * @typedef {Object} TextAreaProps
 * @property {string} initialValue - The initial value of the textarea.
 * @property {function} onChange - Callback function to handle changes in the textarea value.
 */
interface TextAreaProps {
  initialValue: string;
  onChange: (content: string) => void;
}

/**
 * Props for the TextArea2 component.
 * @typedef {Object} TextAreaProps2
 * @property {string} initialValue - The initial value of the textarea.
 * @property {string} Header - The header text to display above the textarea.
 */
interface TextAreaProps2 {
    initialValue: string;
    Header: string;
}

/**
 * TextArea component renders a textarea with an initial value and an onChange handler.
 * @param {TextAreaProps} props - The props for the component.
 * @returns {JSX.Element} The rendered TextArea component.
 */
const TextArea: React.FC<TextAreaProps> = ({ initialValue, onChange }) => {
  const [value, setValue] = useState("");
  useEffect(() => setValue(initialValue), [initialValue]);

  return (
    <textarea
      value={value}
      onChange={(e) => {
        setValue(e.target.value);
        onChange(e.target.value);
      }}
      className="form-control"
      rows={10}
    ></textarea>
  );
};

/**
 * TextArea2 component renders a read-only textarea with a header.
 * @param {TextAreaProps2} props - The props for the component.
 * @returns {JSX.Element} The rendered TextArea2 component.
 */
const TextArea2: React.FC<TextAreaProps2> = ({ initialValue, Header}) => {
    const [value, setValue] = useState("");
    useEffect(() => setValue(initialValue), [initialValue]);

    return (
        <div>
        <Heading headingText={Header}/>
        <textarea
        value={value}
        readOnly
        className="form-control"
        rows={10}
        ></textarea></div>
    );
}

export default TextArea;
export { TextArea2 };