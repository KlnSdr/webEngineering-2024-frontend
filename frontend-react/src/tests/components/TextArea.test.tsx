import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import '@testing-library/jest-dom/extend-expect';
import TextArea from "../../components/TextArea";

describe("TextArea Component", () => {
    test("renders with initial value", () => {
        const initialValue = "Initial Text";
        const handleChange = jest.fn();

        render(<TextArea initialValue={initialValue} onChange={handleChange} />);

        const textarea = screen.getByRole("textbox");
        expect(textarea).toHaveValue(initialValue);
    });

    test("calls onChange callback when value changes", () => {
        const initialValue = "";
        const handleChange = jest.fn();

        render(<TextArea initialValue={initialValue} onChange={handleChange} />);

        const textarea = screen.getByRole("textbox");
        userEvent.type(textarea, "New Value");

        expect(handleChange).toHaveBeenCalledTimes(9); // "New Value" is 9 characters long
        expect(handleChange).toHaveBeenCalledWith("N");
        expect(handleChange).toHaveBeenCalledWith("Ne");
        expect(handleChange).toHaveBeenCalledWith("New");
        expect(handleChange).toHaveBeenCalledWith("New ");
        expect(handleChange).toHaveBeenCalledWith("New V");
        expect(handleChange).toHaveBeenCalledWith("New Va");
        expect(handleChange).toHaveBeenCalledWith("New Val");
        expect(handleChange).toHaveBeenCalledWith("New Valu");
        expect(handleChange).toHaveBeenCalledWith("New Value");
    });

    test("updates value when initialValue prop changes", () => {
        const initialValue = "Initial Text";
        const newValue = "New Initial Text";
        const handleChange = jest.fn();

        const { rerender } = render(<TextArea initialValue={initialValue} onChange={handleChange} />);

        let textarea = screen.getByRole("textbox");
        expect(textarea).toHaveValue(initialValue);

        rerender(<TextArea initialValue={newValue} onChange={handleChange} />);
        textarea = screen.getByRole("textbox");
        expect(textarea).toHaveValue(newValue);
    });
});
