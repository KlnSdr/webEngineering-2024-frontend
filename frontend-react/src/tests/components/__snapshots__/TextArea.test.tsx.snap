// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Component renders with initial value 1`] = `
<DocumentFragment>
  <textarea
    class="form-control"
    rows="10"
  >
    Initial Text
  </textarea>
</DocumentFragment>
`;
